### DISCUSSION TITLE
JWKS endpoint

### DISCUSSION THREAD

**Discussion by caniko** [fastapi-users/fastapi-users#968] [Ideas]
FastAPI-User should also expose a [JSON Web Key Set (JWKS) endpoint](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets) for better integration with other applications.

Here is an example response from a GET request to a JWKS endpoint:
```json
{
    "keys": [
        {
            "kid": "vdaec4Br3ZnRFtZN-pimK9v1eGd3gL2MHu8rQ6M5SiE",
            "kty": "RSA",
            "alg": "RS256",
            "use": "sig",
            "n": "4OPCc_LDhU6ADQj7cEgRei4VUf4PZH8GYsxsR6RSdeKmDvZ48hCSEFiEgfc3FIfh-gC4r9PtKucc_nkRofrAKR4qL8KNNoSuzQAOC92Yz6r7Ao4HppHJ8-QVdo5H-d9wfNSlDLBSo5My4b4EnHb1HLuFxDqyhFpGvsoUJdgbt3m_Q3WAVb2yrM83S6HX__vrQvqUk2e7z5RNrI7LSsW3ZOz9fU7pvm8-kFFAIPJ7fOJIC7UQ9wBWg3YdwQ0B2b24jXjVr0QCGzqJ6o1G_UZYSJCDMGQDpDcEuYnvSKBLfVR-0EcAjolRhcSPjHlW0Cp0YU8qwWDHpjkbrMrFmxlO4Q",
            "e": "AQAB"
        }
    ]
```
`kid` means key id, `alg` is the algorithm, and `n` is the public key used for this realm. Example copied from redhat [article](https://developers.redhat.com/blog/2020/11/24/authentication-and-authorization-using-the-keycloak-rest-api) for keycloak.

The endpoint can be exposed to the public on the discretion of the developer. I will only use it internally.

**Comment by frankie567** [comment]
IMO, I think it's a bit out-of-scope of FastAPI Users.

The JWKS is relevant only when using a JWT strategy with an asymmetric key. Besides, it'll depend greatly of how you manage your key (environment variable? secret stored in a HMS?).

Implementing such endpoint is fairly straightforward, so I guess developers who need it won't have difficulty to implement it. For example, I have [an implementation in my other project](https://github.com/fief-dev/fief/blob/cbfeec11da7a03aa345cb7ceb088b5d8ec9d6ab1/backend/fief/apps/auth/routers/well_known.py#L57-L60), Fief: it's only three lines of code.

```py
@router.get("/jwks.json", name="well_known:jwks")
async def get_jwks(tenant: Tenant = Depends(get_current_tenant)):
    keyset = jwk.JWKSet(keys=tenant.get_sign_jwk())
    return keyset.export(private_keys=False, as_dict=True)
```

**Reply by caniko** [reply]
I don't see how it is out of the scope considering it is authorization based on fast-api users. And, I cannot see a better place to standardize this function than the native package.

Asymmetric key is always better, and should be the standard. The only case against it is more compute (not sure if this is true?) and a small increase in complexity, but in return you get a trustless system for your JWTs.

Moreover, you are correct in saying that it is only a few lines of code. Thanks for sharing. Why not put this in the package, and it could be included a generalised version of the function for it to be included as a router for those that use asymmetric...
